sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -828,21 +829,28 @@ private async Task RunReceiveTaskAsync(CancellationToken cancellationToken)
                             break;
                         }
 
-                        // Do a quick synchronous check before we resort to async/await with the state-machine overhead.
-                        if (!_messageHandlerSemaphore.Wait(0, CancellationToken.None))
+                        bool messageHandlerLockAcquired = false;
+                        try
                         {
-                            try
+                            await _messageHandlerSemaphore.WaitAsync(linkedHandlerTcs.Token).ConfigureAwait(false);
+                            messageHandlerLockAcquired = true;
+                            if (IsSessionProcessor)
                             {
-                                await _messageHandlerSemaphore.WaitAsync(linkedHandlerTcs.Token).ConfigureAwait(false);
+                                await _maxConcurrentAcceptSessionsSemaphore.WaitAsync(linkedHandlerTcs.Token).ConfigureAwait(false);
                             }
-                            catch (OperationCanceledException)
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            if (messageHandlerLockAcquired)
                             {
-                                linkedHandlerTcs.Dispose();
-                                // reset the linkedHandlerTcs if it was already cancelled due to user updating the concurrency
-                                linkedHandlerTcs = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _handlerCts.Token);
-                                // allow the loop to wake up when tcs is signaled
-                                break;
+                                // make sure to release semaphore if we are breaking out of the loop
+                                _messageHandlerSemaphore.Release();
                             }
+                            linkedHandlerTcs.Dispose();
```

Comment:
Should we consider calling `TrySetCancelled` here or do we know that nobody is awaiting?

Reply:
Nobody is awaiting, this loop is single-threaded.